RecoLocalCalo/HGCalRecAlgos/interface/HGCalMultiCluster.h

```diff
@@ -0,0 +1,93 @@
+#ifndef RecoLocalCalo_HGCalRecAlgos_HGCalMultiCluster
+#define RecoLocalCalo_HGCalRecAlgos_HGCalMultiCluster
+
+#include <vector>
+
+#include "DataFormats/EgammaReco/interface/BasicCluster.h"
+
+
+class HGCalMultiCluster{
+
+public:
+  typedef std::vector<reco::BasicCluster>::const_iterator component_iterator;
+  typedef std::vector<reco::BasicCluster> ClusterCollection;
+
+  HGCalMultiCluster(){
+  }
+  HGCalMultiCluster(ClusterCollection &thecls) : myclusters(thecls){
+  }
+  HGCalMultiCluster(const reco::BasicCluster &thecl) :  myclusters(1,thecl){
+  }
+  void push_back(const reco::BasicCluster &b){
+    myclusters.push_back(b);
+  }
+  unsigned int size()
+  {
+    return myclusters.size();
+  }						
+  component_iterator begin() const {
+    return myclusters.begin();
+  }
+  component_iterator end() const {
+    return myclusters.end();
+  }
+  double simple_z(double vz) const {
+    double acc = 0.0;
+    double totweight = 0.;
+    for(component_iterator it = begin(); it != end(); it++){
+      acc += (it->z()-vz)*it->energy()*it->size();
+      totweight+=it->energy()*it->size();
+    }
+    return acc/totweight;
+  }
+
+  double simple_slope_x(double vz) const {
+    double acc = 0.0;
+    double totweight = 0.;
+    for(component_iterator it = begin(); it != end(); it++){
+      acc += it->x()/(it->z()-vz)*it->energy()*it->size();
+      totweight+=it->energy()*it->size();
+    }
+    return acc/totweight;
+  }
+  double simple_slope_y(double vz) const {
+    double acc = 0.0;
+    double totweight = 0.;
+    for(component_iterator it = begin(); it != end(); it++){
+      acc += it->y()/(it->z()-vz)*it->energy()*it->size();
+      totweight+=it->energy()*it->size();
+    }
+    return acc/totweight;
+  }
+  double simple_eta(double vz) const {
+    return -1.*
+      simple_z(vz)/abs(simple_z(vz))*
+      log(tan(atan(sqrt(pow(simple_slope_x(vz),2)+pow(simple_slope_y(vz),2)))/2.));
```

Comment:
@cvuosalo you are absolutely right... due to the various ping-pong of this part of the code an old version got propagated. I will update this area as soon as possible to use a better strategy (though of course as you later point out the tan(atan(something)/2.) is not moot)
All the checks you ask will be added though in some cases I expect it can be proven that the argument cannot take that value (because of other previous checks and the way the argument is calculated. 
By the way thanks for the careful revision of the code...